### PR TITLE
Factor out usage of Guava in shared libraries

### LINKED ISSUE #1412: Use JDK instead of Guava
Guava is a poor choice of a dependency for a library jar because it changes a lot and binary incompatibilities arise frequently downstream. It isn't used all that much in this project so it should be easy to get rid of. There was a comment in #1261 to this effect, but perhaps not enough focus for anyone to take any action.

Example: `io.kubernetes.client.util.ModelMapper` uses a mixture of utilities from Guava and Apache Commons Lang3. Probably it could at least consolidate to only use the Apache library. Ideally it should be possible to settle on just using the JDK.

### PR DESCRIPTION
**Pull request by dsyer**
A few utilities have their own shared implementations
in util package (e.g. Strings, Streams).

ThreadFactory implemented as private method where needed.

Use ConcurrentHashMap instead of Stripe.

The ClassPath scan in ModelMapper is replaced with private
method.

Caffeine is used for caches.

Fixes #1412

### CODE REVIEW

File: examples/pom.xml [hunk 1]
```diff
@@ -17,6 +17,7 @@
 	<name>client-java-examples-parent</name>
 
 	<properties>
+        <guava.version>30.0-jre</guava.version>
```
Comment (yue9944882) [inline examples/pom.xml:20]: nit: fix the indent to align with the line below

File: pom.xml [hunk 2]
```diff
@@ -503,5 +504,51 @@ limitations under the License.
...
       </build>
     </profile>
+    <profile>
+      <id>imports</id>
+      <activation>
+        <os>
+          <family>!windows</family>
+        </os>
+        <file>
+			<missing>.google</missing>
```
Comment (yue9944882) [inline pom.xml:514]: nit: something is wrong with the indents here, is it?

**Review by yue9944882 — COMMENTED** [review]
@dsyer thanks, most changes are trivial. 

a few nits otherwise LGTM. btw why are the following empty files added?

> examples/examples-release-11/.google 
> examples/examples-release-10/.google

File: examples/pom.xml (hunk 1 above)
Comment (dsyer) [reply examples/pom.xml]: The samples use Guava (could also be removed later). These files mark them as "safe" so that the build doesn't fail when `com.google.common` is detected.

**Review by yue9944882 — COMMENTED** [review]
/lgtm

the changes are backward-compatible overall

**Comment by yue9944882** [comment]
@brendandburns for review and approval

File: extended/src/main/java/io/kubernetes/client/extended/controller/Controllers.java [hunk 3]
```diff
@@ -50,6 +51,16 @@
    * @return the thread factory
    */
   public static ThreadFactory namedControllerThreadFactory(String controllerName) {
-    return new ThreadFactoryBuilder().setNameFormat(controllerName + "-%d").build();
+    final ThreadFactory defaultFactory = Executors.defaultThreadFactory();
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/controller/Controllers.java:54]: Please separate out this change as it is unrelated to removing Guava.

File: extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java [hunk 4]
```diff
@@ -334,8 +343,17 @@ private void maybeReportTransition() {
     }
   }
 
-  private ThreadFactory makeThreadFactory(String nameFormat) {
-    return new ThreadFactoryBuilder().setDaemon(true).setNameFormat(nameFormat).build();
+  private ThreadFactory makeThreadFactory(String format) {
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java:346]: Please put this change in a different PR (probably the same PR as the above)

File: pom.xml [hunk 5]
```diff
@@ -503,5 +504,51 @@ limitations under the License.
         </plugins>
       </build>
     </profile>
+    <profile>
```
Comment (brendandburns) [inline pom.xml:507]: Please separate this out into a separate PR

In general we're using Spotless for formatting, and I'm not sure we want to introduce another formatting step.

File: util/src/main/java/io/kubernetes/client/apimachinery/KubernetesRequestDigest.java [hunk 6]
```diff
@@ -102,7 +104,8 @@ public String toString() {
     }
 
     String groupVersion;
-    if (Strings.isNullOrEmpty(resourceMeta.getGroupVersionResource().getGroup())) { // core resource
+    if (Strings.isNullOrEmpty(resourceMeta.getGroupVersionResource().getGroup())) { // core
+      // resource
```
Comment (brendandburns) [inline util/src/main/java/io/kubernetes/client/apimachinery/KubernetesRequestDigest.java:108]: this break is awkward and weird, leave it on the previous line?

File: util/src/main/java/io/kubernetes/client/informer/cache/Controller.java [hunk 7]
```diff
@@ -78,16 +83,25 @@ public Controller(
...
     this.resyncExecutor =
         Executors.newSingleThreadScheduledExecutor(
-            new ThreadFactoryBuilder()
-                .setNameFormat("controller-resync-" + apiTypeClass.getName() + "-%d")
-                .build());
+            threadFactory("controller-reflector-" + apiTypeClass.getName() + "-%d"));
+  }
+
+  private ThreadFactory threadFactory(String format) {
+    final ThreadFactory defaultFactory = Executors.defaultThreadFactory();
```
Comment (brendandburns) [inline util/src/main/java/io/kubernetes/client/informer/cache/Controller.java:95]: same comment as above.

File: util/src/main/java/io/kubernetes/client/util/ModelMapper.java [hunk 8]
```diff
@@ -65,8 +68,9 @@
 
   private static Set<Discovery.APIResource> lastAPIDiscovery = new HashSet<>();
 
-  private static volatile long nextDiscoveryRefreshTimeMillis =
-      0; // zero means api-discovery never done
+  private static volatile long nextDiscoveryRefreshTimeMillis = 0; // zero means
+  // api-discovery
```
Comment (brendandburns) [inline util/src/main/java/io/kubernetes/client/util/ModelMapper.java:72]: this comment breaking is awkward.

File: util/src/main/java/io/kubernetes/client/util/ModelMapper.java [hunk 9]
```diff
@@ -368,43 +378,63 @@ private static void initModelMap() throws IOException {
...
 
   static class BiDirectionalMap<K, V> {
-    private Map<K, V> kvMap = new HashMap<>();
-    private Map<V, K> vkMap = new HashMap<>();
-    private Striped<ReadWriteLock> lock =
-        Striped.readWriteLock(Runtime.getRuntime().availableProcessors() * 4);
+
+    private Map<K, V> kvMap = new ConcurrentHashMap<>();
+
+    private Map<V, K> vkMap = new ConcurrentHashMap<>();
```
Comment (brendandburns) [inline util/src/main/java/io/kubernetes/client/util/ModelMapper.java:384]: What is the locking in here? We used Striped so that there would be multiple locks to reduce contention, I'd rather not go back to a single lock.

**Comment by brendandburns** [comment]
A few comments. I realized after the fact that it's possible that the thread factory changes are because you're removing the Guava thread factory builder and re-implementing what it does.

I still think it's worth separating out those changes into a different PR (esp. because the same code is repeated several times) also, it would be nice to have a pointer to the Guava implementation so that we can validate that the behavior is the same.

Thanks!

**Comment by dsyer** [comment]
The thread factory changes are precisely because we are removing Guava. The original implementation is here: https://github.com/google/guava/blob/master/guava/src/com/google/common/util/concurrent/ThreadFactoryBuilder.java

It's quite hard to separate those changes out, so I'd rather not, unless you have some specific reason to do it. I don't have any problem with tidying up or centralizing the thread factory patterns into a utility if you prefer. That could also happen as a separate PR afterwards.

I'll tidy up the line breaks - it's because my IDE is set to wrap at 80 characters.

File: util/src/main/java/io/kubernetes/client/util/ModelMapper.java (hunk 9 above)
Comment (dsyer) [reply util/src/main/java/io/kubernetes/client/util/ModelMapper.java:384]: `ConcurrentHashMap` has striped locks internally. I didn't see any tests (I know they are hard to write) or comments about the implementation, so I'm not sure what the race condition is that we are trying to avoid. Can you elaborate, and maybe we can improve it?

File: extended/src/main/java/io/kubernetes/client/extended/controller/Controllers.java [hunk 10]
```diff
@@ -50,6 +51,16 @@
...
+    final AtomicInteger threadNumber = new AtomicInteger(1);
+    String format = controllerName + "-%d";
+    return r -> {
+      Thread thread = defaultFactory.newThread(r);
+      if (!thread.isDaemon()) {
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/controller/Controllers.java:59]: Looking at the thread builder code, I don't think that we were previously setting threads to daemon.

File: extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java [hunk 11]
```diff
@@ -334,8 +343,17 @@ private void maybeReportTransition() {
...
+    final ThreadFactory defaultFactory = Executors.defaultThreadFactory();
+    final AtomicInteger threadNumber = new AtomicInteger(1);
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java:348]: This code is repeated, please factor out into a single method that is re-used in various places.

File: util/src/main/java/io/kubernetes/client/informer/cache/Controller.java [hunk 12]
```diff
@@ -78,16 +83,25 @@ public Controller(
...
+    final AtomicInteger threadNumber = new AtomicInteger(1);
```
Comment (brendandburns) [inline util/src/main/java/io/kubernetes/client/informer/cache/Controller.java:96]: here too wrt refactoring to a shared method

File: util/src/main/java/io/kubernetes/client/util/ModelMapper.java (hunk 9 above)
Comment (brendandburns) [reply util/src/main/java/io/kubernetes/client/util/ModelMapper.java:383]: ok, I'm happy if it has striped locks

There's no race condition to avoid, simply that the bi-directional map is used to store API -> type mappings that are used in a number of different places, and we want to reduce contention on that lookup if there are multiple threads all performing lookups at the same time.

**Comment by brendandburns** [comment]
Thanks for the updates!

One comment about the thread behavior changing to always make them daemon threads (I think). We can discuss if we want that change, but it should be a different PR.

One request for a refactor to a shared implementation of thread creation.

Then I think this is read to merge.

File: extended/src/main/java/io/kubernetes/client/extended/controller/Controllers.java (hunk 10 above)
Comment (dsyer) [reply extended/src/main/java/io/kubernetes/client/extended/controller/Controllers.java:59]: Guava was (it's the default). I can't think of any reason you'd want these to be non-daemon.

File: extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java (hunk 11 above)
Comment (dsyer) [reply extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java:348]: It's not repeated in this class. Do you mean you want to have a utility class that is shared across the Java SDK?

Comment (dsyer) [reply extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java]: Did that.

**Review by yue9944882 — COMMENTED** [review]
/lgtm

File: extended/src/main/java/io/kubernetes/client/extended/controller/Controllers.java (hunk 10 above)
Comment (dsyer) [reply extended/src/main/java/io/kubernetes/client/extended/controller/Controllers.java]: I switched it back to non-daemon anyway. I don't know why they weren't daemon before but it does seem to work, so it's safer to leave it as is.

File: pom.xml (hunk 5 above)
Comment (brendandburns) [reply pom.xml:513]: Can you comment on how this is resolved? I don't think these changes are related to Guava.

**Comment by brendandburns** [comment]
Basically LGTM, but I noticed there is one `pom.xml` (formatter?) change that I wanted clarification on.

Thanks!

**Comment by yue9944882** [comment]
> Basically LGTM, but I noticed there is one pom.xml (formatter?) change that I wanted clarification on.

the new plugin forbids guava imports which makes sure we dont step backwards @brendandburns

**Comment by brendandburns** [comment]
Ok, I would rather we did that using a spotless regex (as we did with wildcards) rather than a whole new exec/egrep call.

**Comment by dsyer** [comment]
Do you have an example of how to do that with spotless? I find the user docs a bit thin there.

**Comment by yue9944882** [comment]
@brendandburns i think we can resolve that as a part of https://github.com/kubernetes-client/java/pull/1425, so that we can get this in first. wdyt?

**Comment by dsyer** [comment]
I looked at #1425 for inspiration and I can see what you are doing, but it makes things complicated for now because Guava is still allowed in tests (it ships with Wiremock) and in examples. I'd be happy to add another PR after this one replacing the exec plugin call with another section of spotless config. It won't result in a smaller pom.xml (I predict) but we can see.

**Comment by brendandburns** [comment]
I'd rather just rip out the exec/formatting changes from this PR, and then merge the spotless changes in a follow on PR. I think we can be careful for a few days about re-introducing dependencies.

I don't think the wiremock dependency matters for formatting, since it won't result in any import statements.

And I'm comfortable removing Guava from the examples (I don't think it's super necessary in any of them)

**Comment by dsyer** [comment]
> I'd rather just rip out the exec/formatting changes from this PR

OK. Done. We can deal with the examples and tests when we do the spotless configuration I guess (i.e. another PR)?

**Comment by brendandburns** [comment]
/lgtm
/approve

Thanks for the patience!
